---
title: "SELECTION OF RELEVANT SUBJECTS IN CLINICAL COHORTS WITH OMICS FOR TRANSLATIONAL APPLICATION"
subtitle: "Filtering mechanism"
author: "Fernando Lucas Ruiz (fernando.lucas@um.es)"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    highlight: kate
    number_sections: true
    theme: spacelab
    toc: true
    toc_float: true
    code_folding: "hide"
  pdf_document:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T, warning = F, message = F)
```

# Reading data

Downloading data. I scale the data in the matrix to have homogeneity in the representations.

```{r}
# covariables
ROSMAP_RINPMIAGESEX_covs <- readRDS("~/Library/CloudStorage/OneDrive-UNIVERSIDADDEMURCIA/Documentos/Fernando/Master Bioinformatica/TFM/datos/ROSMAP_RINPMIAGESEX_covs.rds")
covs <- ROSMAP_RINPMIAGESEX_covs
rownames(covs) <- covs$mrna_id
covs$study <- as.factor(covs$study)
covs$projid <- as.character(covs$projid)
covs$ceradsc <- as.factor(covs$ceradsc)
covs$cogdx <- as.factor(covs$cogdx)
covs$neuroStatus <- as.factor(covs$neuroStatus)

# Corrected data
ROSMAP_RINPMIAGESEX_resids <- readRDS("~/Library/CloudStorage/OneDrive-UNIVERSIDADDEMURCIA/Documentos/Fernando/Master Bioinformatica/TFM/datos/ROSMAP_RINPMIAGESEX_resids.rds")
mat_exp <- scale(ROSMAP_RINPMIAGESEX_resids)
```

One-hot encoding of the covariate matrix

```{r}
covs2 <- data.frame(matrix(ncol = 0, nrow = nrow(covs))) 

for (colname in names(covs)) {
  if (is.factor(covs[[colname]]) & length(levels(covs[[colname]])) > 2) {

    dummy_df <- dummy_cols(covs[colname], 
                           remove_selected_columns = TRUE) # quitar las variables iniciales
    
    dummy_df <- data.frame(lapply(dummy_df, factor))

    covs2 <- cbind(covs2, dummy_df)

  } else {

    covs2[[colname]] <- covs[[colname]] # si son numéricas o categóricas binarias, las añadimos igual
  }
}
rownames(covs2) <- covs2$mrna_id
```

# Geneset

```{r}
# geneset de Alzheimer extraidos de KEGG
tab <- getGeneKEGGLinks(species="hsa")
tab$Symbol <- mapIds(org.Hs.eg.db, tab$GeneID,
                       column="SYMBOL", keytype="ENTREZID")

paths <- getKEGGPathwayNames(species="hsa")
geneset_alz <- tab$Symbol[tab$PathwayID=="hsa05010"]
```


# Filtering mechanisms

Filter the matrix Mnxm to Mnxm', with m' << n to filter out predictors, in this case from the expression matrix mat_exp. 

```{r}
mat_exp_alz_genes <- mat_exp[, colnames(mat_exp) %in% geneset_alz]
```

```{r, fig.height=4, fig.width=6}
venn.plot <- venn.diagram(
  x = list(GenesMatriz = colnames(mat_exp), GenesetAlzheimer = geneset_alz),
  category.names = c("Genes of transcriptomic matrix", "Geneset AD\n KEGG"),
  filename = NULL,
  output = FALSE, 
  fill = c("#21908CFF", '#fde725ff'),
  cex = 1,  # Aumenta el tamaño del texto
  fontface = "bold",
  cat.cex = 1, 
  cat.fontface = "bold",
  cat.default.pos = "text",
  cat.pos = c(0, 20), 
  cat.dist = c(0.05, 0.1),
  rotation.degree = 0, 
  margin = 0.1, 
  lwd=0.5,
  lty = "dashed", 
  edge.col = "grey", 
)

grid.newpage() 
grid.draw(venn.plot)
```

```{r}
save.image("filtering mechanism.RData")
```

